lib/ecto/association.ex

```diff
@@ -1396,20 +1655,31 @@ defmodule Ecto.Association.ManyToMany do
     do: changeset
 
   defp field!(op, struct, field) do
-    Map.get(struct, field) || raise "could not #{op} join entry because `#{field}` is nil in #{inspect struct}"
+    Map.get(struct, field) ||
+      raise "could not #{op} join entry because `#{field}` is nil in #{inspect(struct)}"
   end
 
   defp dump!(action, join_through, struct, field, adapter) when is_binary(join_through) do
     value = field!(action, struct, field)
-    type  = struct.__struct__.__schema__(:type, field)
+    type = struct.__struct__.__schema__(:type, field)
 
     case Ecto.Type.adapter_dump(adapter, type, value) do
       {:ok, value} ->
         value
+
       :error ->
-        raise Ecto.ChangeError,
-          "value `#{inspect value}` for `#{inspect struct.__struct__}.#{field}` " <>
-            "in `#{action}` does not match type #{inspect type}"
+        message =
```

Comment:
This is the only meaningful change in this file other than formatting

Reply:
Instead of adding the `case `here, add a function called `Ecto.Type.inspect(type, opts \\ [])` that will be responsible for checking which type it is and delegate accordingly.